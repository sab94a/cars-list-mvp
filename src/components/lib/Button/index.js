// @flow

import React from 'react';
import type { Button as Props} from 'components/lib/types'
import styles from './index.module.scss';

const Button = ({ children, ...rest }: Props) => (
    <button {...rest } className={ styles.root } >{ children }</button>
);

export default React.memo<Props>(Button)
