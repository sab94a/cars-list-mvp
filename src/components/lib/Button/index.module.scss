@import 'components/lib/variables.scss';

.root {
    border: 0;
    background: $primary;
    color: #fff;
    width: $button-width;
    font-size: $font-size-sm;
    padding: $offset-sm;
    border-radius: 2px;
    cursor: pointer;

    &:hover, &:focus {
        background: $primary-pressed;
    }
}