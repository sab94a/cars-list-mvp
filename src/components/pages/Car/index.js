//@flow

import React from 'react';
import styles from './index.module.scss';

const Car = () => (<div className={styles.root}>Car</div>)

export default Car;
