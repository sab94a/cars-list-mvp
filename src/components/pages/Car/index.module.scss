@import 'components/lib/variables.scss';

.root {
    width: 100%;
}

.image {
    min-height: 345px;
    background-color: $light;
    background-position: center;
    background-size: contain;
    background-repeat: no-repeat;

    &.hasImage {
        background-color: transparent;
    }
}

.wrapper {
    display: flex;
    align-items: flex-start;
    max-width: 800px;
    margin: 0 auto;
    padding: $offset-lg 0;
}

.content {
    flex-grow: 1;
    padding-right: 2 * $offset-lg;
}

.aside {
    width: $offset-lg * 2 + $button-width * 2;
    flex-shrink: 0;
    padding: $offset-lg;
    border: $border-width-sm solid $light;
    text-align: right;

    & > p {
        margin: 0;
        font-size: $font-size-sm;
        text-align: left;
        line-height: 1.7;
    }
}

.title {
    margin: 0;
    margin-bottom: $offset-lg;
    font-size: $font-size-lg;
}

.description {
    font-size: $font-size-md;
    margin-bottom: $offset-lg;
}

.note {
    margin: 0;
    font-size: $font-size-sm;
    line-height: 1.8;
}
