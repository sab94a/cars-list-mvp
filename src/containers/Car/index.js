// @flow

import CarPage from 'components/pages/Car';
import { connect } from 'react-redux';

import type { CarView } from 'types/views';
import type { ReduxState, Dispatch } from 'types/store';
import type { RouterProps } from 'types/route';

import { selectCar, selectCarError } from 'selectors';
import { initCar, addFavourite, removeFavourite } from 'actions';

export type StateProps = {
    car: ?CarView,
    error: ?string,
    isFavourite: boolean
};

export type dispatchProps = {
    init: () => mixed,
    addFavourite: () => mixed,
    removeFavourite: () => mixed
};

const mapStateToProps = (state:ReduxState, props: RouterProps) => ({
    car: selectCar(state, props),
    error: selectCarError(state),
});

const mapDispatchToProps = (dispatch: Dispatch, { match: { params } }:RouterProps):dispatchProps => ({
    init: () => dispatch(initCar(+params.id)),
    addFavourite: () => dispatch(addFavourite(+params.id)),
    removeFavourite: () => dispatch(removeFavourite(+params.id)),
});

export default connect(mapStateToProps, mapDispatchToProps)(CarPage);
