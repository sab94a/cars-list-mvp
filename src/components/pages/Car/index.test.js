import React from 'react';
import { mount } from 'enzyme';
import toJson from 'enzyme-to-json';

import Button from 'components/lib/Button';

import Car from './';

describe('<Car /> page component', () => {
    let init;
    let update;
    let addFavourite;
    let removeFavourite;
    let navigate;

    const car = {
        stockNumber: 1,
        title: 'My Super Car',
        image: 'image_car.png',
        description: 'description',
        isFavourite: false
    };

    let defaultProps;
    let component;

    const create = params => {
        component = mount(<Car { ...params } />);
    };

    beforeEach(() => {
        init = jest.fn();
        update = jest.fn();
        addFavourite = jest.fn();
        removeFavourite = jest.fn();
        navigate = jest.fn();

        defaultProps = {
            init,
            update,
            addFavourite,
            removeFavourite,
            navigate,
            car
        }
    });

    it('Should render correctly', () => {
        create({ ...defaultProps });

        expect(toJson(component)).toMatchSnapshot();
    });

    it('Should call init after mount if car is absent', () => {
        create({ ...defaultProps, car: null }); 
        expect(init).toHaveBeenCalled();
    });

    it('Should NOT call init after mount if car is absent', () => {
        create({ ...defaultProps }); 
        expect(init).not.toHaveBeenCalled();
    });

    it('Should render banner with correct image', () => {
        const $image = component.find('.image');

        expect($image.prop('style')).toEqual({ backgroundImage: `url(${ car.image })`})
    });

    it('Should render correct title', () => {
        const $title = component.find('h1');
            
        expect($title.text()).toBe(car.title);
    });

    it('Should render correct description', () => {
        const $description = component.find('.description');
            
        expect($description.text()).toBe(car.description);
    });

    it('Should render Save Button if car is not Favourite', () => {
        create({ ...defaultProps });

        const $btn = component.find(Button);
        const cb = $btn.prop('onClick');

        cb();

        expect($btn.text()).toBe('Save');
        expect(addFavourite).toHaveBeenCalled();
    });

    it('Should render Remove Button if car is Favourite', () => {
        create({ ...defaultProps, car: { ...car, isFavourite: true } });
        
        const $btn = component.find(Button);
        const cb = $btn.prop('onClick');

        cb();

        expect($btn.text()).toBe('Remove');
        expect(removeFavourite).toHaveBeenCalled();
    });

    it('Should navigate to 404 if error was passed', () => {
        component.setProps({ ...defaultProps, error: 'error', car: null })

        expect(navigate).toHaveBeenCalled();
        expect(navigate).toHaveBeenCalledWith('/404');
    })
});
