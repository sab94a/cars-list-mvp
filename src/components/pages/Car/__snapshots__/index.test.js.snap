// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Car /> page component Should render correctly 1`] = `
<Car
  addFavourite={[MockFunction]}
  car={
    Object {
      "description": "description",
      "image": "image_car.png",
      "isFavourite": false,
      "stockNumber": 1,
      "title": "My Super Car",
    }
  }
  init={[MockFunction]}
  navigate={[MockFunction]}
  removeFavourite={[MockFunction]}
  update={[MockFunction]}
>
  <div
    className="root"
  >
    <div
      className="image hasImage"
      style={
        Object {
          "backgroundImage": "url(image_car.png)",
        }
      }
    />
    <div
      className="wrapper"
    >
      <div
        className="content"
      >
        <h1
          className="title"
        >
          My Super Car
        </h1>
        <p
          className="description"
        >
          description
        </p>
        <p
          className="note"
        >
          This car is currently available and can be delivered as soon as tomorrow morning. Please be aware that delivery times shown in this page are not definitive and may change due to bad weather conditions.
        </p>
      </div>
      <aside
        className="aside"
      >
        <p>
          If you like this car, click the button and save it in your collection of favourite items.
        </p>
        <Memo(Button)
          onClick={[MockFunction]}
        >
          <button
            className="root"
            onClick={[MockFunction]}
          >
            Save
          </button>
        </Memo(Button)>
      </aside>
    </div>
  </div>
</Car>
`;
